File: viewer/src/utils/renderer.test.ts
import { describe, expect, it } from "bun:test";
import { formatRendererSpec, parseRendererSpec, type RendererSpec } from "./renderer";

describe("parseRendererSpec", () => {
  it.each<[string, RendererSpec]>([
    [
      "owner/repo/name@1.2.3",
      { owner: "owner", repo: "repo", name: "name", version: "1.2.3" },
    ],
    [
      "owner/repo/name",
      { owner: "owner", repo: "repo", name: "name", version: "latest" },
    ],
    [
      "my.org/my_repo/my-name@v1.0.0-beta",
      { owner: "my.org", repo: "my_repo", name: "my-name", version: "v1.0.0-beta" },
    ],
  ])("parses valid spec %s", (input, expected) => {
    expect(parseRendererSpec(input)).toEqual(expected);
  });

  it.each([
    ["path traversal in owner",    "../evil/repo/name"],
    ["path traversal in repo",     "owner/../evil/name"],
    ["path traversal in name",     "owner/repo/../../etc"],
    ["slash injection in version", "owner/repo/name@v1/evil"],
    ["spaces in a segment",        "ow ner/repo/name"],
    ["percent-encoded characters", "owner/re%2Fpo/name"],
    ["empty owner",                "/repo/name"],
    ["empty repo",                 "owner//name"],
    ["empty name",                 "owner/repo/"],
    ["empty version",              "owner/repo/name@"],
    ["too few segments",           "owner/repo"],
    ["too many segments",          "owner/repo/name/extra"],
    ["empty string",               ""],
  ])("rejects %s", (_label, input) => {
    expect(parseRendererSpec(input)).toBeNull();
  });
});

describe("formatRendererSpec", () => {
  it.each<[string, RendererSpec]>([
    [
      "owner/repo/name@1.2.3",
      { owner: "owner", repo: "repo", name: "name", version: "1.2.3" },
    ],
    [
      "owner/repo/name",
      { owner: "owner", repo: "repo", name: "name", version: "latest" },
    ],
    [
      "my.org/my_repo/my-name@v1.0.0-beta",
      { owner: "my.org", repo: "my_repo", name: "my-name", version: "v1.0.0-beta" },
    ],
  ])("formats spec as %s", (expected, spec) => {
    expect(formatRendererSpec(spec)).toBe(expected);
  });
});

File: mobile/test/renderer_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:txtshr/renderer.dart';

void main() {
  group('RendererSpec.parse', () {
    test('parses owner/repo/name without version', () {
      final spec = RendererSpec.parse('alice/my-repo/markdown');
      expect(spec, isNotNull);
      expect(spec!.owner, 'alice');
      expect(spec.repo, 'my-repo');
      expect(spec.name, 'markdown');
      expect(spec.version, 'latest');
    });

    test('parses owner/repo/name@version', () {
      final spec = RendererSpec.parse('alice/my-repo/markdown@1.2.3');
      expect(spec, isNotNull);
      expect(spec!.version, '1.2.3');
    });

    test('returns null for too few segments', () {
      expect(RendererSpec.parse('alice/my-repo'), isNull);
    });

    test('returns null for too many segments', () {
      expect(RendererSpec.parse('alice/my-repo/name/extra'), isNull);
    });

    test('returns null for unsafe characters in segment', () {
      expect(RendererSpec.parse('ali ce/repo/name'), isNull);
      expect(RendererSpec.parse('alice/repo/na/me'), isNull);
      expect(RendererSpec.parse('alice/repo/name@bad version'), isNull);
    });

    test('returns null for empty string', () {
      expect(RendererSpec.parse(''), isNull);
    });
  });

  group('RendererSpec.toString', () {
    test('omits version when latest', () {
      final spec = RendererSpec.parse('alice/repo/name');
      expect(spec.toString(), 'alice/repo/name');
    });

    test('includes version when not latest', () {
      final spec = RendererSpec.parse('alice/repo/name@2.0.0');
      expect(spec.toString(), 'alice/repo/name@2.0.0');
    });
  });
}
